### ISSUE TITLE
Gradients not flowing with backward

### ISSUE THREAD

**Issue by legohyl** [rusty1s/pytorch_geometric#980]
## ❓ Questions & Help

Hi guys, I'd like to ask some questions regarding gradient flows. It appears like when I do my loss.backward() function, the parameters don't get updated. I'm trying to use GCNConv to learn node representations, and build a classifier on top of these representations. Ultimately, I wish to train all of this end to end, but it looks like the gradients don't flow backwards and my training loss is just cyclical. Here's a code snippet on what I've tried to do.

```
class AssignmentGCN(Module):
    def __init__(self, d_feat_len, p_feat_len, repr_size, gcn_size, num_embedding, d_idx, p_idx):
        super(AssignmentGCN, self).__init__()
        self.d_feat_len = d_feat_len
        self.p_feat_len = p_feat_len
        self.d_gh_idx = d_idx
        self.p_gh_idx = p_idx

        # Shared Embedding
        self.geohash_embed = nn.Embedding(num_embeddings=num_embedding, embedding_dim=repr_size)
        self.input_size = repr_size + d_feat_len - 1 # embedding + feats
        
        self.gcn1 = GCNConv(self.input_size, 4*gcn_size, normalize=False)
        self.gcn2 = GCNConv(4*gcn_size, gcn_size, normalize=False)
    
        self.lin = nn.Linear((gcn_size*2)+1, 1)
        self.act_lin = nn.Sigmoid()
        
        self.reset_params()
        
    def forward(self, graph_feats, graph_gh, edge_index, edge_weight):

        # Shared Embedding
        graph_gh_embed = self.geohash_embed.forward(graph_gh)
        graph_gh_embed = graph_gh_embed.view(graph_feats.size(0), -1)

        graph_feats = torch.cat((graph_feats, graph_gh_embed), dim=1)

        x = self.gcn1.forward(graph_feats, edge_index=edge_index, edge_weight=edge_weight)
        x = F.relu(x)
        x = self.gcn2(x, edge_index=edge_index, edge_weight=edge_weight)

        
        # Take the d p vectors, and concat them, followed by putting classifier on top 
        stop_num = int(edge_index.shape[1]/2)
        
        node_idx = edge_index.T[0:stop_num]
        
        d_graph_rep = x[node_idx[:,0]]
        p_graph_rep = x[node_idx[:,1]]
        edge_weight_half = edge_weight[0:stop_num].type(torch.FloatTensor).view(-1,1).cuda()
        
        d_p_pair = torch.cat((d_graph_rep, p_graph_rep, edge_weight_half), dim=1)
        
        x = self.lin(d_p_pair)
        y = self.act_lin(x)
        return y
```

**Comment by rusty1s** [comment]
Gradients are not flowing since you call the modules via the `forward` function. Simply removing this should fix this.

**Comment by legohyl** [comment]
Thanks for spotting that! I correct the code and removed the `forward` functions. However, it still looks like the gradients are not flowing. Would the last part of the code cause the problem? Does re-arranging the data cause the gradients to break? 
```
        # Take the d p vectors, and concat them, followed by putting classifier on top 
        stop_num = int(edge_index.shape[1]/2)
        
        node_idx = edge_index.T[0:stop_num]
        
        d_graph_rep = x[node_idx[:,0]]
        p_graph_rep = x[node_idx[:,1]]
        edge_weight_half = edge_weight[0:stop_num].type(torch.FloatTensor).view(-1,1).cuda()
        
        d_p_pair = torch.cat((d_graph_rep, p_graph_rep, edge_weight_half), dim=1)
        
        x = self.lin(d_p_pair)
        y = self.act_lin(x)
```

**Comment by rusty1s** [comment]
This shouldn't break anything. It all looks differentiable to me. It would be best to look into where the gradient is lost exactly.
BTW: `edge_index.T[0:stop_num]` doesn't give you the upper triangle of the sparse adjacency matrix, it gives you the upper half. You can fix this via:
```
row, col = edge_index
mask = row < col
undirected_edge_index = edge_index[:, mask]
```

**Comment by legohyl** [comment]
Hmm, if I replace the entire chunk of the linear layer and just take the outputs of the GCN and take some L1Loss, all the parameters in the 2 GCN layers and the embedding layer get updated. So it appears like that last step causes the gradients to stop.

As for your suggestion, if i'm not wrong, the `edge_index` is a `2 x N` tensor which is indicating `from` and `to` right? I wanted to get the edge indexes such that I know which rows of the node representations to subset and concat together to form the pair. So `x` is basically of length `d_nodes + p_nodes`, where `d_nodes` are connected to some `p_nodes` and vice versa. I wish to build a classifier on top of a concat of `(d_node, p_node, edge_weight)`. Hence I subset the `x` accordingly to get the representations and concat them on `dim=1`.

**Comment by legohyl** [comment]
So I replaced my way of subsetting the edges to the mask solution you provided, and I've also changed out all the forward functions, but it still appears like the gradients aren't flowing. In order to make them flow, I tried not combining the node features together. Here's a snippet of the model now. If I avoid subsetting the graph features and just use some arbitrary label, it looks like the loss.backward() works. Hence, it appears that subsetting the features then concatenating them causes the gradients to stop?

```
# GCN Model
class AssignmentGCN(Module):
    def __init__(self, dafeat_len, p_feat_len, repr_size, gcn_size, num_embedding, d_idx, p_idx):
        super(AssignmentGCN, self).__init__()
        self.d_feat_len = d_feat_len
        self.p_feat_len = p_feat_len
        self.d_gh_idx = d_idx
        self.p_gh_idx = p_idx

        # Shared Embedding
        self.geohash_embed = nn.Embedding(num_embeddings=num_embedding, embedding_dim=repr_size)
        self.input_size = repr_size + d_feat_len - 1 # embedding + feats
        
        self.gcn1 = GCNConv(self.input_size, 4*gcn_size, normalize=False)
        self.gcn2 = GCNConv(4*gcn_size, gcn_size, normalize=False)
    
        self.lin = nn.Linear(gcn_size*1, 1)
        self.act_lin = nn.Sigmoid()
        
        self.reset_params()
        
    def forward(self, graph_feats, graph_gh, edge_index, edge_weight):

        # Shared Embedding
        # Grab the column for GH embedding & concat with the rest of the features
        graph_gh_embed = self.geohash_embed(graph_gh)
        graph_gh_embed = graph_gh_embed.view(graph_feats.size(0), -1)
    
        # Combine in the embedded geohash
        graph_feats = torch.cat((graph_feats, graph_gh_embed), dim=1)

        x = self.gcn1(graph_feats, edge_index=edge_index, edge_weight=edge_weight)
        x = F.relu(x)
        x = self.gcn2(x, edge_index=edge_index, edge_weight=edge_weight)

        # Take the dax pax vectors, and concat them, followed by putting classifier on top 
        stop_num = int(edge_index.shape[1]/2)
        row, col = edge_index
        mask = row < col
        node_idx = edge_index[:, mask]
        dax_graph_rep = x[node_idx[0,:], :]
        pax_graph_rep = x[node_idx[1,:], :]
        
        dax_pax_pair = x
        x = self.lin(dax_pax_pair)
        y = self.act_lin(x)

        return y
```

**Comment by rusty1s** [comment]
This is super weird. Indexing shouldn't stop the gradients from flowing. Is there some form of error if you do this or does your model simply stops to train?

**Comment by legohyl** [comment]
There's no errors which is puzzling. I backpropagate the errors and do `torch.sum(model.lin.weights)` to see if the parameters are changing, and they seem to be flat from the start. At the same time, I plot my losses over time, and it's just cyclical in nature.

**Comment by legohyl** [comment]
I'm beginning to think that it's likely due to my data set? Would that cause any issue? Even then, it doesn't explain why (1) parameters are fixed in value (linear layer doesn't get any updates as the sum of all it's parameters don't change) and (2) my loss plot is basically cyclical and each time it sees the same graph, the same loss value is achieved.

**Comment by legohyl** [comment]
So I brought the data processing out of the training loop, just to iterate purely on one graph. It doesn't seem to help at all. I compute the linear layer weights and they're all the same, the loss values are all the same on a straight line. I even increased the learning rate to >1 so that the optimizer takes a huge step but that doesn't seem to change anything.

```
graph_num = d_nodes.window_num.max()
losses = []
start_time = time.time()
model.train()

gn = 5
small_set_d = d_nodes.loc[d_nodes.window_num == gn].copy().reset_index(drop=True)
small_set_p = p_nodes.loc[p_nodes.window_num == gn].copy().reset_index(drop=True)
gfeat, gh, edge_in, edge_wt, lbls = build_p_d_graph(small_set_d, small_set_p)
gfeat = gfeat.cuda()
edge_in = edge_in.cuda()
edge_wt = edge_wt.cuda()
gh = gh.cuda()
lbls = lbls.cuda()

for epoch in range(20):
    for gn in range(1):
        optimizer.zero_grad()
        out  = model(gfeat, gh, edge_in, edge_wt)

        fake_lab = torch.ones((int(edge_in.size(1)/2), 1)).cuda()
        loss = criterion(out, fake_lab)
        loss.backward()
        optimizer.step()
        losses.append(loss.data.cpu().numpy())
        print(torch.sum(model.lin.weight))
end_time = time.time()
```

**Comment by legohyl** [comment]
I think I found the issue. After tinkering around, I realized the nn.Sigmoid layer was the problem. It seems like the BCELoss was the issue. If I swapped to the BCEWithLogitsLoss, it seems like the loss behaves and the weight matrices actually change now! I'm not really sure why BCELoss causes this issue though, and whether it's the numerical instability that Pytorch indicates. Nevertheless, I think it solves the mystery?
